File: query_funcs.cpp
#include "query_funcs.h"

// return number of tuples in "tableName"
int countNum(connection *C, string tableName) {
	pqxx::work w(*C);
	pqxx::result r = w.exec("SELECT COUNT(*) FROM \"" + tableName + "\";");
	w.commit();
	return r[0][0].as<int>();
}

void add_player(connection *C, int team_id, int jersey_num, string first_name, string last_name,
                int mpg, int ppg, int rpg, int apg, double spg, double bpg)
{
  //	int playerID = countNum(C, "PLAYER") + 1;
	pqxx::work w(*C);
	// where is player id ?
	pqxx::result r = w.exec("INSERT INTO \"PLAYER\" (\"TEAM_ID\", \"UNIFORM_NUM\", \"FIRST_NAME\", \"LAST_NAME\", \"MPG\", \"PPG\", \"RPG\", \"APG\", \"SPG\", \"BPG\") VALUES("
	                        + to_string(team_id) + ","
	                        + to_string(jersey_num) + ",'" + first_name + "','" + last_name + "',"
	                        + to_string(mpg) + "," + to_string(ppg) + "," + to_string(rpg) + ","
	                        + to_string(apg) + "," + to_string(spg) + "," + to_string(bpg) + ");");
	w.commit();
}


void add_team(connection *C, string name, int state_id, int color_id, int wins, int losses)
{
  //	int teamID = countNum(C, "TEAM") + 1;
	pqxx::work w(*C);
	pqxx::result r = w.exec("INSERT INTO \"TEAM\" (\"NAME\", \"STATE_ID\", \"COLOR_ID\", \"WINS\", \"LOSSES\") VALUES('"
	                        + name + "'," + to_string(state_id) + "," + to_string(color_id) + ","
	                        + to_string(wins) + "," + to_string(losses) + ");");
	w.commit();
}


void add_state(connection *C, string name)
{
  //	int stateID = countNum(C, "STATE") + 1;
	pqxx::work w(*C);
	pqxx::result r = w.exec("INSERT INTO \"STATE\" (\"NAME\") VALUES('"
	                        + name + "');");
	w.commit();
}


void add_color(connection *C, string name)
{
  //	int colorID = countNum(C, "COLOR") + 1;
	pqxx::work w(*C);
	pqxx::result r = w.exec("INSERT INTO \"COLOR\" (\"NAME\") VALUES('"
	                        + name + "');");
	w.commit();
}


void query1(connection *C,
            int use_mpg, int min_mpg, int max_mpg,
            int use_ppg, int min_ppg, int max_ppg,
            int use_rpg, int min_rpg, int max_rpg,
            int use_apg, int min_apg, int max_apg,
            int use_spg, double min_spg, double max_spg,
            int use_bpg, double min_bpg, double max_bpg
           )
{
	string cmd;
	if (use_mpg == 0 && use_ppg == 0 && use_rpg == 0 && use_apg == 0 && use_spg == 0 && use_bpg == 0) {
		cmd = "SELECT * FROM \"PLAYER\"";
	}
	else {
		cmd = "SELECT * FROM \"PLAYER\" WHERE";
		cmd = use_mpg == 1 ? cmd + " \"MPG\">=" + to_string(min_mpg) + " AND \"MPG\" <=" + to_string(max_mpg) + " AND" : cmd;
		cmd = use_ppg == 1 ? cmd + " \"PPG\">=" + to_string(min_ppg) + " AND \"PPG\" <=" + to_string(max_ppg) + " AND" : cmd;
		cmd = use_rpg == 1 ? cmd + " \"RPG\">=" + to_string(min_rpg) + " AND \"RPG\" <=" + to_string(max_rpg) + " AND" : cmd;
		cmd = use_apg == 1 ? cmd + " \"APG\">=" + to_string(min_apg) + " AND \"APG\" <=" + to_string(max_apg) + " AND" : cmd;
		cmd = use_spg == 1 ? cmd + " \"SPG\">=" + to_string(min_spg) + " AND \"SPG\" <=" + to_string(max_spg) + " AND" : cmd;
		cmd = use_bpg == 1 ? cmd + " \"BPG\">=" + to_string(min_bpg) + " AND \"BPG\" <=" + to_string(max_bpg) + " AND" : cmd;
		if(cmd[cmd.size()-1 == 'D']) {
			cmd = cmd.substr(0, cmd.size() - 4);
		}
		cmd += ";";
	}

	cout << "PLAYER_ID TEAM_ID UNIFORM_NUM FIRST_NAME LAST_NAME MPG PPG RPG APG SPG BPG" << endl;
	pqxx::work w(*C);
	pqxx::result r = w.exec(cmd.c_str());
	w.commit();
	for(int i = 0; i < r.size(); i++) {
		for(int j = 0; j < 11; j++) {
			cout << r[i][j].as<string>();
			if(j != 10) {cout << " ";}
		}
		cout << endl;
	}
}


void query2(connection *C, string team_color)
{
	string cmd = "SELECT \"TEAM\".\"NAME\" FROM \"TEAM\", \"COLOR\" WHERE \"TEAM\".\"COLOR_ID\"=\"COLOR\".\"COLOR_ID\" AND \"COLOR\".\"NAME\"='" + team_color + "';";
	pqxx::work w(*C);
	pqxx::result r = w.exec(cmd.c_str());
	w.commit();
	cout << "NAME" << endl;
	for(int i = 0; i < r.size(); i++) {
		cout << r[i][0] << endl;
	}
}


void query3(connection *C, string team_name)
{
	string cmd = "SELECT \"PLAYER\".\"FIRST_NAME\", \"PLAYER\".\"LAST_NAME\" FROM \"PLAYER\", \"TEAM\" WHERE \"PLAYER\".\"TEAM_ID\"=\"TEAM\".\"TEAM_ID\" AND \"TEAM\".\"NAME\"='" + team_name + "' ORDER BY \"PLAYER\".\"PPG\" DESC;";
	pqxx::work w(*C);
	pqxx::result r = w.exec(cmd.c_str());
	w.commit();
	cout << "FIRST_NAME LAST_NAME" << endl;
	for(int i = 0; i < r.size(); i++) {
		cout << r[i][0] << " " << r[i][1] << endl;
	}
}


void query4(connection *C, string team_state, string team_color)
{
	string cmd = "SELECT \"FIRST_NAME\", \"LAST_NAME\", \"UNIFORM_NUM\" FROM \"PLAYER\", \"TEAM\", \"STATE\", \"COLOR\" WHERE \"PLAYER\".\"TEAM_ID\"=\"TEAM\".\"TEAM_ID\" AND \"TEAM\".\"STATE_ID\"=\"STATE\".\"STATE_ID\" AND \"TEAM\".\"COLOR_ID\"=\"COLOR\".\"COLOR_ID\" AND \"STATE\".\"NAME\"='" + team_state + "' AND \"COLOR\".\"NAME\"='" + team_color + "';";
	pqxx::work w(*C);
	pqxx::result r = w.exec(cmd.c_str());
	w.commit();
	cout << "FIRST_NAME LAST_NAME UNIFORM_NUM" << endl;
	for(int i = 0; i < r.size(); i++) {
		cout << r[i][0] << " " << r[i][1] << " " << r[i][2] << endl;
	}
}


void query5(connection *C, int num_wins)
{
	string cmd = "SELECT \"FIRST_NAME\", \"LAST_NAME\", \"TEAM\".\"NAME\", \"TEAM\".\"WINS\" FROM \"PLAYER\", \"TEAM\" WHERE \"PLAYER\".\"TEAM_ID\"=\"TEAM\".\"TEAM_ID\" AND \"TEAM\".\"WINS\" > " + to_string(num_wins) + ";";
	pqxx::work w(*C);
	pqxx::result r = w.exec(cmd.c_str());
	w.commit();
	cout << "FIRST_NAME LAST_NAME NAME WINS" << endl;
	for(int i = 0; i < r.size(); i++) {
		cout << r[i][0] << " " << r[i][1] << " " << r[i][2] << " " << r[i][3] << endl;
	}
}

File: extra/ORM/query_funcs.py
import os
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "ORM.settings")
import django
django.setup()
from ormApp.models import *

def add_player(team_id, jersey_num, first_name, last_name,
             mpg, ppg, rpg, apg, spg, bpg):
    PLAYER.objects.get_or_create(TEAM_ID=TEAM.objects.get(TEAM_ID=team_id), UNIFORM_NUM=jersey_num, FIRST_NAME=first_name,
                          LAST_NAME=last_name, MPG=mpg, PPG=ppg, RPG=rpg, APG=apg, SPG=spg, BPG=bpg)

def add_team(name, state_id, color_id, wins, losses):
    TEAM.objects.get_or_create(NAME=name, STATE_ID=STATE.objects.get(STATE_ID=state_id),
                        COLOR_ID=COLOR.objects.get(COLOR_ID=color_id), WINS=wins, LOSSES=losses)

def add_state(name):
    STATE.objects.get_or_create(NAME=name)

def add_color(name):
    COLOR.objects.get_or_create(NAME=name)

def query1(use_mpg, min_mpg, max_mpg,
             use_ppg, min_ppg, max_ppg,
             use_rpg, min_rpg, max_rpg,
             use_apg, min_apg, max_apg,
             use_spg, min_spg, max_spg,
             use_bpg, min_bpg, max_bpg):
    print("PLAYER_ID TEAM_ID UNIFORM_NUM FIRST_NAME LAST_NAME MPG PPG RPG APG SPG BPG")
    res = PLAYER.objects.all()
    if use_mpg == 1:
        res = res.filter(MPG__gte=min_mpg, MPG__lte=max_mpg)
    if use_ppg == 1:
        res = res.filter(PPG__gte=min_ppg, PPG__lte=max_ppg)
    if use_rpg == 1:
        res = res.filter(RPG__gte=min_rpg, RPG__lte=max_rpg)
    if use_apg == 1:
        res = res.filter(APG__gte=min_apg, APG__lte=max_apg)
    if use_spg == 1:
        res = res.filter(SPG__gte=min_spg, SPG__lte=max_spg)
    if use_bpg == 1:
        res = res.filter(BPG__gte=min_bpg, BPG__lte=max_bpg)
    for ply in res:
        print(str(ply.PLAYER_ID) + " " + str(ply.TEAM_ID.TEAM_ID) + " " + str(ply.UNIFORM_NUM) + " " + str(ply.FIRST_NAME) + " "
              + str(ply.LAST_NAME) + " " + str(ply.MPG) + " " + str(ply.PPG) + " " + str(ply.RPG) + " " + str(ply.APG)
              + " " + str(ply.SPG) + " " + str(ply.BPG))

def query2(team_color):
    print('NAME')
    res = TEAM.objects.all()
    res = res.filter(COLOR_ID__NAME=team_color)
    for team in res:
        print(team.NAME)

def query3(team_name):
    print('FIRST_NAME LAST_NAME')
    res = PLAYER.objects.all()
    res = res.filter(TEAM_ID__NAME=team_name).order_by('-PPG')
    for ply in res:
        print(ply.FIRST_NAME + ' ' + ply.LAST_NAME)

def query4(team_state, team_color):
    print('FIRST_NAME LAST_NAME UNIFORM_NUM')
    res = PLAYER.objects.all();
    res = res.filter(TEAM_ID__STATE_ID__NAME=team_state, TEAM_ID__COLOR_ID__NAME=team_color)
    for ply in res:
        print(ply.FIRST_NAME + ' ' + ply.LAST_NAME + ' ' + str(ply.UNIFORM_NUM))

def query5(num_wins):
    print('FIRST_NAME LAST_NAME NAME WINS')
    res = PLAYER.objects.all()
    res = res.filter(TEAM_ID__WINS__gt=num_wins)
    for ply in res:
        print(ply.FIRST_NAME + ' ' + ply.LAST_NAME + ' ' + ply.TEAM_ID.NAME + ' ' + str(ply.TEAM_ID.WINS))

query1(1, 35, 40,
    0, 10, 20,
    0, 1, 5,
    0, 1, 10,
    0, 0.3, 1,
    0, 0.2, 1)
print('*************************************************************')
query2("Maroon")
print('*************************************************************')

query3("Duke")
print('*************************************************************')

query4("NC", "DarkBlue")
print('*************************************************************')

query5(10)
